# Part 1. Equation of a Slime

How many late days are you using for this assignment? 0

In [116]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
import numpy as np
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

## 1. Loading the dataset

In [117]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

df = pd.read_csv('science_data_large.csv')
print(df.head(15))
print(df.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [118]:
# Take the pandas dataset and split it into our features (X) and label (y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 

X = df[['Temperature °C', 'Mols KCL']]  # Features
y = df['Size nm^3']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
print("Training feature set shape (X_train):", X_train.shape)
print("Testing feature set shape (X_test):", X_test.shape)
print("Training label set shape (y_train):", y_train.shape)
print("Testing label set shape (y_test):", y_test.shape)

Training feature set shape (X_train): (900, 2)
Testing feature set shape (X_test): (100, 2)
Training label set shape (y_train): (900,)
Testing label set shape (y_test): (100,)


## 3. Perform a Linear Regression

In [119]:
# Use sklearn to train a model on the training set

# Create a sample datapoint and predict the output of that sample with the trained model

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX

model = LinearRegression()
model.fit(X_train, y_train)

sample_data = pd.DataFrame([[500, 600]], columns=['Temperature °C', 'Mols KCL'])  
predicted_size = model.predict(sample_data)

print("Predicted size for the sample data point [Temperature = 500°C, Mols KCl = 600]:", predicted_size[0])

train_score = model.score(X_train, y_train)
print("Training score (R²):", train_score)

coefficients = model.coef_
intercept = model.intercept_

print("Coefficients:", coefficients)
print("Intercept:", intercept)

Predicted size for the sample data point [Temperature = 500°C, Mols KCl = 600]: 643298.7669940306
Training score (R²): 0.8610082946223788
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340764


Write the linear equation of a slime: (example equation: $E = mc^2$)

The equation for our linear regression model is:

$$ h(x) = \theta_0 + \theta_1 \cdot \text{Temperature °C} + \theta_2 \cdot \text{Mols KCL} $$

Where:
- \(\theta_0\) is the intercept
- \(\theta_1\) is the coefficient for Temperature °C
- \(\theta_2\) is the coefficient for Mols KCL

The equation for the slime growth model is:

$$ h(x) = -409391.47958340764 + 866.14641337 \cdot \text{Temperature °C} + 1032.69506649 \cdot \text{Mols KCL} $$

Report on score and explain meaning:

The score reported by the Linear Regression model is the coefficient of determination, commonly known as R². This value ranges from 0 to 1 (though it can be negative with very poor fits), where:

- 1 indicates perfect prediction where the model explains all the variability in the data
- 0 means the model fails to explain any of the variability (as good as just predicting the mean)
- A negative score (possible with test data) means the model performs worse than just predicting the mean

In simpler terms, it measures how well our linear equation fits the actual data points. A higher score means our predictions are closer to the real values

In this case, the R² score of the model is approximately **0.8610**. This means that the model explains approximately **86.10%** of the variance in slime size based on the features provided (Temperature and Mols of KCl).

The R² score of **0.8610** indicates a good fit for the model, suggesting that it effectively captures the relationship between the input features and the slime growth size. Further validation with unseen data is recommended to confirm the model's generalizability.

## 4. Use Cross Validation

In [120]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42

# Report on their finding and their significance

model = LinearRegression()
scores = cross_val_score(model, X, y, cv=5)

print("Cross-validation scores for each fold:", scores)
print("Mean cross-validation score:", scores.mean())
print("Standard deviation of cross-validation scores:", scores.std())


Cross-validation scores for each fold: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Mean cross-validation score: 0.8568167899144437
Standard deviation of cross-validation scores: 0.01346630737209602


Write findings here:

We conducted a 5-fold cross-validation on our linear regression model using the `cross_val_score` function from scikit-learn. This process divided our dataset into 5 subsets, training the model on 4 folds and testing on the remaining 1 fold, iterating this process 5 times with different test folds.

#### Findings:
- **Individual Fold Scores**: The R² scores for each of the 5 folds were: [0.8392, 0.8705, 0.8587, 0.8720, 0.8436]
- **Mean R² Score**: The average R² score across all folds was 0.8568
- **Standard Deviation**: The standard deviation of the R² scores was 0.0135

#### Significance:


1. **Model Performance**: The mean R² score of 0.8568 indicates that the model, on average, explains approximately 85.68% of the variance in the size of the slime growth based on the input features (temperature and mols of KCl). A score close to 1 indicates that our model performs well in generalizing to unseen data, suggesting a robust relationship between the input features and the growth size of the slime.

2. **Standard Deviation**: The calculated standard deviation of the cross-validation scores was 0.0135. This low standard deviation implies that the model's performance is both consistent and stable across different data shuffles, highlighting that our model’s performance does not vary significantly between different training and testing splits.
3. **Comparison to Single Split**: In Part 3, we obtained a training R² score of 0.8610 from a single 90/10 train-test split. The cross-validation mean R² of 0.8568 is very close to this value (a difference of only 0.0042, or 0.42%), which strongly supports the reliability of our original model. This small difference suggests that the model is not overfitting to the single split from Part 3 and performs similarly well across multiple data shuffles. The cross-validation score, being an average over multiple tests, is likely a more realistic estimate of how the model would perform on unseen data.

## 5. Using Polynomial Regression

In [121]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)

# Report on the metrics and output the resultant equation as you did in Part 3.

# Create polynomial features of degree 2
degree = 2
polynomial_features = PolynomialFeatures(degree=degree)

X_poly_train = polynomial_features.fit_transform(X_train)
X_poly_test = polynomial_features.fit_transform(X_test)

model_poly = LinearRegression()
model_poly.fit(X_poly_train, y_train)

poly_score = model_poly.score(X_poly_test, y_test)
print("Polynomial regression score (R²):", poly_score)

coefficients = model_poly.coef_
intercept = model_poly.intercept_

print("Coefficients:", coefficients)
print("Intercept:", intercept)
print(f"Polynomial Equation: h(x) = {coefficients[0]} + {coefficients[1]} * Temperature + {coefficients[2]} * Mols KCl + {coefficients[3]} * Temperature^2 + {coefficients[4]} * Temperature * Mols KCl + {coefficients[5]} * Mols KCl^2" + f" + {intercept}")

# Use the PolynomialFeatures library to repeat the experiment for degrees 1-10 and k-fold cross-validation

cv_scores_poly = cross_val_score(model_poly, X_poly_train, y_train, cv=5)

print("Cross-validation scores for each fold:", cv_scores_poly)
print("Mean cross-validation score:", cv_scores_poly.mean())
print("Standard deviation of cross-validation scores:", cv_scores_poly.std())




Polynomial regression score (R²): 1.0
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27197088e-07  1.26458843e-11
  2.00000000e+00  2.85714287e-02]
Intercept: 2.0479375962167978e-05
Polynomial Equation: h(x) = 0.0 + 11.999999977691196 * Temperature + -1.2719708848869804e-07 * Mols KCl + 1.2645884339690383e-11 * Temperature^2 + 2.0000000000172227 * Temperature * Mols KCl + 0.028571428708642932 * Mols KCl^2 + 2.0479375962167978e-05
Cross-validation scores for each fold: [1. 1. 1. 1. 1.]
Mean cross-validation score: 1.0
Standard deviation of cross-validation scores: 0.0


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

h(x) = 11.999999977691196 * `Temperature` +
       1.2719708848869804e-07 * `Mols KCl` + 
       1.2645884339690383e-11 * `Temperature^2` +
       2.0000000000172227 * `Temperature` * `Mols KCl` + 
       0.028571428708642932 * `Mols KCl^2` + 
       2.0479375962167978e-05







Report on the score and interpret:

The polynomial regression model achieved a perfect R² score of **1.0**, indicating that it explains **100%** of the variance in the size of the slime based on the features provided (Temperature and Mols of KCl). 

The cross-validation results also confirm this finding, as every fold during validation scored **1.0**, showing that the model maintains high performance consistently across different subsets of the data. 

The standard deviation of **0.0** reinforces that there is no variability in the model's performance, suggesting that it has perfectly learned the underlying relationships in the training data.

While such a perfect score is impressive, it may also indicate overfitting, meaning that the model could be too closely fitting the training data and may not generalize well to unseen data. 

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [122]:
# Load the dataset. Then train and evaluate the classification models.

df = pd.read_csv('ckd_feature_subset.csv')
df.head()

,age,bp,wbcc,appet_poor,appet_good,rbcc,Target_ckd
0,0.688312,0.333333,0.000000,1,0,0.000000,1
1,0.545455,0.333333,0.128319,1,0,0.305085,1
2,0.714286,0.500000,0.238938,1,0,0.186441,1
3,0.688312,0.333333,0.283186,0,1,0.338983,1
4,0.441558,0.333333,0.221239,1,0,0.220339,1


In [123]:
X = df.drop(columns='Target_ckd')  # Feature variables
y = df['Target_ckd']                # Target variable

# Check the shapes to confirm the split
print(f"Features shape: {X.shape}")
print(f"Target shape: {y.shape}")

Features shape: (153, 6)
Target shape: (153,)


In [124]:
# Initialize models
models = {
    'Logistic Regression': LogisticRegression(max_iter=1000, random_state=42),
    'Support Vector Machine': SVC(random_state=42),
    'k-Nearest Neighbors': KNeighborsClassifier(),
    'Neural Network': MLPClassifier(max_iter=1000, random_state=42)
}

# Dictionary to hold results
results = {}

# Evaluate each model using 5-fold cross-validation
for name, model in models.items():
    scores = cross_val_score(model, X, y, cv=5)
    results[name] = {
        'Mean Accuracy': np.mean(scores),
        'Standard Deviation': np.std(scores)
    }

# Convert results into a DataFrame for better readability
results_df = pd.DataFrame(results).T
results_df

,Mean Accuracy,Standard Deviation
Logistic Regression,0.856129,0.054236
Support Vector Machine,0.928172,0.059993
k-Nearest Neighbors,0.928172,0.031893
Neural Network,0.941290,0.024040


In [125]:
# Determine which model performed best
best_model = results_df['Mean Accuracy'].idxmax()
best_score = results_df['Mean Accuracy'].max()

print(f"The best model is: {best_model} with an average accuracy of {best_score:.4f}.")

The best model is: Neural Network with an average accuracy of 0.9413.


## Key Observations

1. **Neural Network Performance**:
   - The Neural Network achieved the highest mean accuracy of **0.9413**, indicating that it correctly classified approximately 94.13% of CKD cases during cross-validation.
   - It also had the lowest standard deviation of **0.0240**, suggesting robust and consistent performance across different folds.

2. **Support Vector Machine and k-Nearest Neighbors**:
   - Both SVM and KNN achieved a mean accuracy of **0.9282**. This strong performance shows that both models effectively classify CKD cases.
   - However, KNN displayed a lower standard deviation of **0.0319** compared to SVM's **0.0599**, indicating slightly more consistency in its results.

3. **Logistic Regression**:
   - The Logistic Regression model had the lowest mean accuracy of **0.8561**.
   - Its standard deviation of **0.0542** indicates more variability in its performance compared to the Neural Network and KNN.

## Summary of Results

The results indicate that the **Neural Network** was the top performer among the tested models. It not only achieved the highest accuracy but also demonstrated the most consistent results across different data splits.

I think the neural network performed the best as neural networks are capable of modeling complex, non-linear relationships between the input features and the target variable. In the context of the CKD dataset, the interactions between various clinical features may not be linearly separable, and a Neural Network can effectively learn these intricate relationships.

I also think another major factor for its performance is the Hierarchical Feature Learning architecture. Neural Networks operate by layering neurons in multiple layers (hidden layers). This architecture allows them to learn increasingly abstract representations of the input data. For instance, initial layers may detect simple features (like ranges of age or blood pressure), while deeper layers can identify complex patterns indicative of CKD.

Logistic regression and k nearest neighbors performed the worst and this could be since they are not as good as capturing non-linear relationships and patterns with many features involved.

## Results and Conclusion for Classification Experiments

In [126]:
# Experiments with Neural Network.

# Different configurations for Neural Networks
best_configs = {
    'Config 1': MLPClassifier(hidden_layer_sizes=(20, 20), 
                              max_iter=2000, 
                              alpha=0.01,  # L2 regularization
                              random_state=42),
    
    'Config 2': MLPClassifier(hidden_layer_sizes=(10,), 
                              max_iter=2000, 
                              learning_rate_init=0.01, 
                              activation='tanh', 
                              random_state=42),
    
    'Config 3': MLPClassifier(hidden_layer_sizes=(10,), 
                              max_iter=2000, 
                              alpha=0.001,  # Smaller L2 regularization
                              learning_rate_init=0.001, 
                              random_state=42)
}

# Dictionary to hold results for the best configurations
best_results = {}

# Evaluate each configuration using 5-fold cross-validation
for config_name, model in best_configs.items():
    scores = cross_val_score(model, X, y, cv=5)
    best_results[config_name] = {
        'Mean Accuracy': np.mean(scores),
        'Standard Deviation': np.std(scores)
    }

# Convert results into a DataFrame for better readability
best_results_df = pd.DataFrame(best_results).T
best_results_df

,Mean Accuracy,Standard Deviation
Config 1,0.967312,0.029342
Config 2,0.973763,0.032140
Config 3,0.921505,0.044574


In [ ]:
# Determine which neural network configuration performed best
best_nn_model = best_results_df['Mean Accuracy'].idxmax()
best_nn_score = best_results_df['Mean Accuracy'].max()
print(f"The best neural network configuration is: {best_nn_model} with an average accuracy of {best_nn_score:.4f}.")


The best neural network configuration is: Config 2 with an average accuracy of 0.9738.


## Results and Conclusion for Neural Network Experiments


1. **Configuration 2 (Increased Learning Rate)**:
   - **Mean Accuracy**: Achieved the highest mean accuracy of **0.9738**. This indicates a successful learning process that maximally captures the underlying patterns in the data.
   - **Standard Deviation**: At **0.0321**, the model showed slight variability, but overall performance remained robust.

2. **Configuration 1 (Two Hidden Layers)**:
   - **Mean Accuracy**: Followed closely with a mean accuracy of **0.9673**. 
   - **Standard Deviation**: A standard deviation of **0.0293**, indicating consistent performance.

3. **Configuration 3 (L2 Regularization)**:
   - **Mean Accuracy**: The lowest performance with a mean accuracy of **0.9215**.
   - **Standard Deviation**: A higher standard deviation of **0.0446** suggests more variability in its results across the folds.

### Conclusion on Parameters Impact

1. **Key Parameter: Learning Rate**:
   - The learning rate appeared to greatly influence the model's performance, as Configuration 2 with an increased learning rate of **0.01** resulted in the highest accuracy.

2. **Hidden Layer Architecture**:
   - Configuration 1, which incorporated two hidden layers, also performed strongly, suggesting that more complex structures can effectively learn intricate patterns in the CKD dataset.

3. **Impact of Regularization**:
   - Configuration 3, despite using L2 regularization, had the lowest accuracy. This suggests that while regularization is useful in preventing overfitting, it may not be optimal in all configurations, particularly with lower learning rates.

Based on the results, the learning rate and number of hidden layers appear to have the most substantial impact on the model's classification performance. An increase in either or both tends to enhance the model's accuracy. This suggests that a higher learning rate may accelerate the convergence of the model during training, allowing it to learn more quickly and effectively from the data.